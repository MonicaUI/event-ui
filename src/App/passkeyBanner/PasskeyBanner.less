.wrapper {
  padding: 0.5rem 0.75rem 0.75rem;
  display: flex;

  .iconWrapper {
    font-size: 1.25rem;

    .icon {
      font-size: 1.875em;
      margin: 2px;
    }
  }

  .contentWrapper {
    flex: 1;
    margin-left: 0.75rem;
  }

  .header {
    margin: 0.25rem 0 0.5rem;
  }
}

.actionBar {
  margin-top: 0.75rem;
}
