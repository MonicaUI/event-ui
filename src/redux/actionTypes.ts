export const LOAD_ACCOUNT_SNAPSHOT = 'event-guestside-site/LOAD_ACCOUNT_SNAPSHOT';
export const LOAD_EVENT_SNAPSHOT = 'event-guestside-site/LOAD_EVENT_SNAPSHOT';
export const LOAD_WEBSITE_CONTENT = 'event-guestside-site/LOAD_WEBSITE_CONTENT';
export const SET_REFERRER = 'event-guestside-site/SET_REFERRER';
export const LOAD_REGISTRATION_PROCESS_CONTENT = 'event-guestside-site/LOAD_REGISTRATION_PROCESS_SNAPSHOT';
export const LOAD_VISIBLE_PRODUCTS = 'event-guestside-site/LOAD_VISIBLE_PRODUCTS';
export const LOAD_REG_CART_VISIBLE_PRODUCTS = 'event-guestside-site/LOAD_REG_CART_VISIBLE_PRODUCTS';
export const REPLACE_VISIBLE_SESSION_PRODUCTS = 'event-guestside-site/REPLACE_VISIBLE_SESSION_PRODUCTS';
export const LOAD_VISIBLE_SESSION_PRODUCTS = 'event-guestside-site/LOAD_VISIBLE_SESSION_PRODUCTS';
export const LOAD_VISIBLE_PRODUCTS_BY_REG_TYPE = 'event-guestside-site/LOAD_VISIBLE_PRODUCTS_BY_REG_TYPE';
export const LOAD_EVENT_WEB_APP_SETTINGS = 'event-guestside-site/LOAD_EVENT_WEB_APP_SETTINGS';
export const SET_WEBSITE_PASSWORD_VALIDITY = 'event-guestside-site/SET_WEBSITE_PASSWORD_VALIDITY';
export const UPDATE_QUESTION_VISIBILTY = 'event-guestside-site/UPDATE_QUESTION_VISIBILITY';
export const GET_ORDERS = 'event-guestside-site/GET_ORDERS';
export const CLEAR_ORDERS = 'event-guestside-site/CLEAR_ORDERS';
export const SWITCH_CCPA_DIALOG_MODE = 'event-guestside-site/SWITCH_CCPA_DIALOG_MODE';
export const SHOW_TRANSPARENT_WRAPPER = 'event-guestside-site/SHOW_TRANSPARENT_WRAPPER';
export const HIDE_TRANSPARENT_WRAPPER = 'event-guestside-site/HIDE_TRANSPARENT_WRAPPER';
export const SPINNER_SELECTION_PENDING = 'event-guestside-site/SPINNER_SELECTION_PENDING';
export const SPINNER_SELECTION_DONE = 'event-guestside-site/SPINNER_SELECTION_DONE';
export const RESTORE_ALL_REG_TYPES_FOR_EVENT = 'event-guestside-site/RESTORE_ALL_REG_TYPES_FOR_EVENT';
