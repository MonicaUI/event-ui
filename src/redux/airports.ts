import { getIn } from 'icepick';
import { getAirActualsFromBooking } from './travelCart/airActuals';
import { ALL_AIR_ACTUALS_SOURCES } from 'event-widgets/utils/travelConstants';
import baseReducer, {
  getAirportIdsFromSnapshot,
  loadAirports,
  setAirports
} from 'event-widgets/redux/modules/airports';
export { setAirports };

const SEARCH_AIRPORTS_SUCCESS = 'event-guestside-site/airports/SEARCH_AIRPORTS_SUCCESS';

const MIN_LENGTH_FOR_SEARCH = 3;

export function setAirportSearch(searchText: $TSFixMe, airports: $TSFixMe): $TSFixMe {
  return { type: SEARCH_AIRPORTS_SUCCESS, payload: { searchText, airports } };
}

const getAirportIdsFromTravelCart = state => {
  const bookings = getIn(state, ['travelCart', 'cart', 'bookings']);
  const airportsToFetch = [];
  // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
  if (bookings && bookings.length) {
    bookings.forEach(b => {
      // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
      if (b.airBookings && b.airBookings.length) {
        b.airBookings.forEach(ab => {
          if (ab.departureFrom) {
            airportsToFetch.push(ab.departureFrom);
          }
          if (ab.departureTo) {
            airportsToFetch.push(ab.departureTo);
          }
        });
      }
      const allAirActuals = getAirActualsFromBooking(b, ALL_AIR_ACTUALS_SOURCES);
      if (allAirActuals.length) {
        allAirActuals.forEach(airActual => {
          // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
          if (airActual.flightDetails && airActual.flightDetails.length) {
            airActual.flightDetails.forEach(fd => {
              if (fd.departureFrom) {
                airportsToFetch.push(fd.departureFrom);
              }
              if (fd.arrivalTo) {
                airportsToFetch.push(fd.arrivalTo);
              }
            });
          }
        });
      }
    });
  }
  return airportsToFetch;
};

/**
 * Creates a thunked action to load the required airports for an existing travel cart.
 */
export const loadAirportsForTravelCart =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const airportsToFetch = getAirportIdsFromTravelCart(getState());
    if (airportsToFetch.length) {
      await dispatch(loadAirports(airportsToFetch));
    }
  };

/**
 * Creates a thunked action to load the required airports for an event.
 */
export const loadAirportsForEvent =
  (eventSnapshotVersion: $TSFixMe, travelSnapshotVersion: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventSnapshotClient }
    } = getState();
    const { airports } = await eventSnapshotClient.getEventTravelAirports(eventSnapshotVersion, travelSnapshotVersion);
    dispatch(setAirports(airports));
  };

/**
 * Creates a thunked action to load the required airports for an event and the loaded travel cart
 */
export const loadAirportsForEventAndTravelCart =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const airportsFromSnapshot = getAirportIdsFromSnapshot(state);
    const airportsFromTravelCart = getAirportIdsFromTravelCart(state);
    const allAirportsToFetch = [...airportsFromSnapshot, ...airportsFromTravelCart];
    if (allAirportsToFetch.length) {
      await dispatch(loadAirports(allAirportsToFetch));
    }
  };

/**
 * Creates a thunked action to load airports by search text.
 * @param searchText The text to search for.
 */
export const loadAirportsForSuggestions =
  (searchText: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { lookupClient }
    } = getState();
    if (searchText && searchText.length >= MIN_LENGTH_FOR_SEARCH) {
      const response = await lookupClient.searchAirports(searchText);
      const { airports } = response;
      dispatch(setAirportSearch(searchText, airports));
    }
  };

const initialState = {
  searchResults: {}
};

export default function reducer(state = initialState, action: $TSFixMe): $TSFixMe {
  switch (action.type) {
    case SEARCH_AIRPORTS_SUCCESS: {
      return {
        ...state,
        ...action.payload.airports,
        searchResults: {
          ...state.searchResults,
          [action.payload.searchText]: action.payload.airports
        }
      };
    }
    default: {
      return baseReducer(state, action);
    }
  }
}
