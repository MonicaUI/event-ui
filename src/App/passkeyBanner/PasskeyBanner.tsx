import React from 'react';
import { connect } from 'react-redux';
import StickyPageBanner from '../StickyPageBanner';
import PasskeyBannerContent from './PasskeyBannerContent';
import { getActivePasskeyBookings } from '../../redux/selectors/currentRegistrant';
import { getCurrentPageId } from '../../redux/pathInfo';
import { modifyPasskeyRequest } from '../../redux/travelCart/passkeyHotelRequest';
import { CANCELLATION } from '../../redux/website/registrationProcesses';

// eslint-disable-next-line react/prop-types
const PasskeyBanner = ({ translate, isCancellationPage, passkeyBookings, onCancelRequest, style }) => {
  // eslint-disable-next-line react/prop-types
  const showPasskeyBanner = isCancellationPage && passkeyBookings.length > 0;
  return showPasskeyBanner ? (
    // @ts-expect-error ts-migrate(2769) FIXME: No overload matches this call.
    <StickyPageBanner>
      <PasskeyBannerContent
        translate={translate}
        passkeyBookings={passkeyBookings}
        actionButtons={[
          {
            name: 'cancel',
            text: 'EventWidgets_Passkey_RegCancelStickyCancelButton__resx',
            callback: onCancelRequest,
            isPrimary: true
          }
        ]}
        icon="attentionWarning"
        contentDetails={<span>{translate('EventWidgets_Passkey_RegCancelStickyFirstLine__resx')}</span>}
        style={style}
        title="EventWidgets_Passkey_RegCancelStickyHeader__resx"
      />
    </StickyPageBanner>
  ) : null;
};

const getBannerStyle = globalTheme => {
  return {
    ...globalTheme,
    header: { styleMapping: 'text1' },
    content: { styleMapping: 'body1' },
    primaryButton: { styleMapping: 'primaryButton' },
    secondaryButton: { styleMapping: 'secondaryButton' }
  };
};

const mapStateToProps = (state: $TSFixMe) => {
  const { global } = state.website.theme;
  return {
    translate: state.text.translate,
    passkeyBookings: getActivePasskeyBookings(state),
    isCancellationPage: CANCELLATION.isTypeOfPage(state, getCurrentPageId(state)),
    style: getBannerStyle(global)
  };
};

const mapDispatchToProps = {
  onCancelRequest: modifyPasskeyRequest
};

export default connect(mapStateToProps, mapDispatchToProps)(PasskeyBanner);
