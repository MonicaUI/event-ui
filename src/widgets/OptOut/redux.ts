/* eslint-disable import/no-cycle */

import { injectPageSnapshot } from '../../redux/actions';
import querystring from 'querystring';

/**
 * Opt-out and opt-in Redux module:
 * Event invitees can click "Opt Out" link in their emails to opt them out all communications from the event account
 */
const GET_OPT_OUT_STATUS = 'event-guestside-site/optOut/GET_OPT_OUT_STATUS';
const OPT_OUT_CONTACT = 'event-guestside-site/optOut/OPT_OUT_CONTACT';

const createGetOptedOutStatusAction = contactInfo => {
  return {
    type: GET_OPT_OUT_STATUS,
    payload: contactInfo
  };
};

export const getOptOutStatus =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventGuestClient },
      accessToken,
      event: { id: eventId }
    } = getState();
    let {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    if (contactId) {
      inviteeId = null;
    }
    const optedOutContact = await eventGuestClient.getOptOutStatus(accessToken, eventId, inviteeId, guestId, contactId);
    dispatch(createGetOptedOutStatusAction(optedOutContact));
  };

export const createOptedOutAction = (contactInfo: $TSFixMe): $TSFixMe => {
  return {
    type: OPT_OUT_CONTACT,
    payload: contactInfo
  };
};

export const optOutAction =
  (optOutStatus: $TSFixMe, fromConfirmPage: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventGuestClient },
      accessToken,
      event: { id: eventId }
    } = getState();
    let {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    if (contactId) {
      inviteeId = null;
    }
    // eslint-disable-next-line import/no-cycle
    const page = (await import(/* webpackChunkName optOut */ './content')).optInOutPageNew;
    if ((fromConfirmPage && optOutStatus) || !fromConfirmPage) {
      const optedOutContact = await eventGuestClient.optOut(
        accessToken,
        eventId,
        inviteeId,
        guestId,
        contactId,
        optOutStatus
      );
      dispatch(createOptedOutAction(optedOutContact));
    }
    dispatch(injectPageSnapshot(page));
  };

export const prepareForOptOutPageLoad =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    let page;
    if (!inviteeId && !guestId && !contactId) {
      page = (await import(/* webpackChunkName optOut */ './invalidInviteeContent')).invalidInviteeOptOut;
    } else {
      await dispatch(getOptOutStatus());
      if (getState() && getState().optOut && getState().optOut.optOutStatus) {
        // eslint-disable-next-line import/no-cycle
        page = (await import(/* webpackChunkName optOut */ './content')).optInOutPage;
      } else {
        // eslint-disable-next-line import/no-cycle
        page = (await import(/* webpackChunkName optOut */ './content')).confirmationPage;
      }
    }
    dispatch(injectPageSnapshot(page));
  };

export const optOutReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case GET_OPT_OUT_STATUS:
    case OPT_OUT_CONTACT:
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
