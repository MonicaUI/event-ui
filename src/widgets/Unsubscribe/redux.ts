import { injectPageSnapshot } from '../../redux/actions';
import querystring from 'querystring';

/**
 * Unsubscribe and re-subscribe Redux module:
 * Event invitees can click "Unsubscribe" link in their emails to opt them out from the event emails communication
 */
const GET_UNSUBSCRIBE_STATUS = 'event-guestside-site/unsubscribe/GET_UNSUBSCRIBE_STATUS';
const SET_UNSUBSCRIBE_STATUS = 'event-guestside-site/unsubscribe/SET_UNSUBSCRIBE_STATUS';

const createGetUnsubscribeStatusAction = (unsubscribedContact, optOutStatus) => {
  return {
    type: GET_UNSUBSCRIBE_STATUS,
    payload: {
      ...optOutStatus,
      unsubscribeStatus: !unsubscribedContact.subscribed
    }
  };
};

export const getUnsubscribeStatus =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventEmailClient, eventGuestClient },
      accessToken,
      event: { id: eventId },
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    const unsubscribedContact = await eventEmailClient.getSubscriptionStatus(eventId, inviteeId, encodedGuestId);
    // optOutStatus object has invitee/guest and planner infomation for display
    const optOutStatus = await eventGuestClient.getOptOutStatus(accessToken, eventId, inviteeId, encodedGuestId);
    dispatch(createGetUnsubscribeStatusAction(unsubscribedContact, optOutStatus));
  };

const createUnsubscribeAction = unsubscribeStatus => {
  return {
    type: SET_UNSUBSCRIBE_STATUS,
    payload: { unsubscribeStatus }
  };
};

export const unsubscribeAction =
  (unsubscribeStatus: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventEmailClient },
      event: { id: eventId },
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    const subscribed = !unsubscribeStatus;
    await eventEmailClient.setSubscriptionStatus(eventId, inviteeId, encodedGuestId, subscribed);
    dispatch(createUnsubscribeAction(unsubscribeStatus));
  };

export const prepareForUnsubscribePageLoad =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    let unsubscribePage;
    if (!inviteeId && !encodedGuestId) {
      unsubscribePage = (await import(/* webpackChunkName: "unsubscribe" */ './invalidInviteeContent'))
        .invalidInviteeUnsubscribePage;
    } else {
      await dispatch(getUnsubscribeStatus());
      // eslint-disable-next-line import/no-cycle
      unsubscribePage = (await import(/* webpackChunkName: "unsubscribe" */ './content')).unsubscribePage;
    }
    dispatch(injectPageSnapshot(unsubscribePage));
  };

export const unsubscribeReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case GET_UNSUBSCRIBE_STATUS:
    case SET_UNSUBSCRIBE_STATUS:
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
