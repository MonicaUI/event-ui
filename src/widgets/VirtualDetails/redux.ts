import querystring from 'querystring';
import { injectPageSnapshot } from '../../redux/actions';
import { populateAllProducts } from '../../redux/visibleProducts';
import { isBeforeStartTime, currentTimeIsInRangeOf } from 'event-widgets/redux/modules/timezones';
import { CANCELLED } from 'event-widgets/clients/EventStatus';
import { sessionStatus } from 'event-widgets/redux/selectors/event';
import Logger from '@cvent/nucleus-logging';

const SET_VIRTUAL_DETAILS = 'event-guestside-site/virtualDetails/SET_VIRTUAL_DETAILS';
const DISPLAY_SESSION_NOT_AVAILABLE_MESSAGE =
  'event-guestside-site/virtualDetails/DISPLAY_SESSION_NOT_AVAILABLE_MESSAGE';
export const EVENT = 'event';
export const SESSION = 'session';
export const VIRTUAL_DETAILS_PAGE = 'virtualDetailsPage';
export const SESSION_NOT_AVAILABLE_PAGE = 'sessionNotAvailableMessagePage';
const LOG = new Logger('widgets/virtualDetails/redux');

const setVirtualDetails = ({ code, url, type, isPassed, pageToDisplay }) => {
  return {
    type: SET_VIRTUAL_DETAILS,
    payload: { code, url, type, isPassed, pageToDisplay }
  };
};

const displaySessionNotAvailableMessage = () => {
  return {
    type: DISPLAY_SESSION_NOT_AVAILABLE_MESSAGE,
    payload: { pageToDisplay: SESSION_NOT_AVAILABLE_PAGE }
  };
};

export const joinWebcastSession =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      virtualDetails: { url }
    } = getState();
    if (url) {
      window.open(url, '_blank');
    }
  };

const getURLAndCode = (attendeeWebcastData, genericWebcastData) => {
  let url;
  let code;

  if (attendeeWebcastData?.join?.href) {
    url = attendeeWebcastData?.join?.href;
    code = attendeeWebcastData?.join?.code;
  } else if (genericWebcastData?.joiningURL) {
    url = genericWebcastData?.joiningURL;
    code = genericWebcastData?.joiningCode;
  }

  return { url, code };
};

const fetchSessionWebcastData = (attendeeId, sessionId) => async (dispatch, getState) => {
  let {
    // eslint-disable-next-line prefer-const
    selectedTimeZone,
    // eslint-disable-next-line prefer-const
    text: { eventTimezone },
    // eslint-disable-next-line prefer-const
    clients: { universalWebcastClient },
    visibleProducts: { Widget: { sessionProducts = {} } = {} }
  } = getState();
  let virtualData = {};
  let attendeeLinksContainer;
  let attendeeWebcastData;

  await dispatch(populateAllProducts('Widget'));
  sessionProducts = getState().visibleProducts?.Widget?.sessionProducts;
  const selectedSessionDetails = sessionProducts[sessionId];

  // If session was deleted in test mode
  if (!selectedSessionDetails) {
    dispatch(displaySessionNotAvailableMessage());
    return;
  }

  if (selectedSessionDetails.status === sessionStatus.cancelled) {
    return;
  }

  const isSessionPassed = !(
    currentTimeIsInRangeOf(
      selectedSessionDetails.startTime,
      selectedSessionDetails.endTime,
      selectedTimeZone,
      eventTimezone
    ) || isBeforeStartTime(selectedSessionDetails.startTime, selectedTimeZone, eventTimezone)
  );

  if (isSessionPassed) {
    if (selectedSessionDetails?.webcast?.recordingURL) {
      virtualData = {
        url: selectedSessionDetails?.webcast?.recordingURL,
        code: selectedSessionDetails?.webcast.recordingCode
      };
    } else {
      return;
    }
  } else {
    try {
      attendeeLinksContainer = await universalWebcastClient.getAttendeeLinkData(attendeeId);
    } catch (error) {
      LOG.warn(`Failed to get attendee specific session webcast data for attendeeId ${attendeeId}`);
      return;
    }
    attendeeWebcastData = attendeeLinksContainer?.data.find(
      attendeeLinkData => attendeeLinkData?.session?.id === sessionId
    );
    virtualData = getURLAndCode(attendeeWebcastData, selectedSessionDetails.webcast);
  }

  if (!(virtualData as $TSFixMe)?.url) {
    return;
  }

  dispatch(
    // @ts-expect-error ts-migrate(2345) FIXME: Argument of type '{ type: string; isPassed: boolea... Remove this comment to see the full error message
    setVirtualDetails({ ...virtualData, type: SESSION, isPassed: isSessionPassed, pageToDisplay: VIRTUAL_DETAILS_PAGE })
  );
};

const fetchEventWebcastData = attendeeId => async (dispatch, getState) => {
  const {
    event: { status: eventStatus, webcast: eventWebcast, startDate, endDate },
    selectedTimeZone,
    text: { eventTimezone },
    clients: { universalWebcastClient }
  } = getState();
  let virtualData = {};
  let attendeeLinksContainer;
  let attendeeWebcastData;

  if (eventStatus === CANCELLED) {
    return;
  }

  const isEventPassed = !(
    currentTimeIsInRangeOf(startDate, endDate, selectedTimeZone, eventTimezone) ||
    isBeforeStartTime(startDate, selectedTimeZone, eventTimezone)
  );

  if (isEventPassed) {
    if (eventWebcast?.recordingURL) {
      virtualData = {
        url: eventWebcast?.recordingURL,
        code: eventWebcast?.recordingCode
      };
    } else {
      return;
    }
  } else {
    try {
      attendeeLinksContainer = attendeeId ? await universalWebcastClient.getAttendeeLinkData(attendeeId) : undefined;
    } catch (error) {
      LOG.warn(`Failed to get attendee specific event webcast data for attendeeId ${attendeeId}`);
      return;
    }
    attendeeWebcastData = attendeeLinksContainer?.data.find(
      attendeeLinkData => Object.keys(attendeeLinkData?.session).length === 0
    );
    virtualData = getURLAndCode(attendeeWebcastData, eventWebcast);
  }

  if (!(virtualData as $TSFixMe)?.url) {
    return;
  }

  dispatch(
    // @ts-expect-error ts-migrate(2345) FIXME: Argument of type '{ type: string; isPassed: boolea... Remove this comment to see the full error message
    setVirtualDetails({ ...virtualData, type: EVENT, isPassed: isEventPassed, pageToDisplay: VIRTUAL_DETAILS_PAGE })
  );
};

export const prepareForVirtualDetailsPageLoad =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const sessionId = queryParams.sessionId;
    const attendeeId = guestId || inviteeId;
    let virtualDetailsPage;

    if (sessionId) {
      if (attendeeId) {
        await dispatch(fetchSessionWebcastData(attendeeId, sessionId));
      }
    } else {
      await dispatch(fetchEventWebcastData(attendeeId));
    }

    if (getState()?.virtualDetails?.pageToDisplay === VIRTUAL_DETAILS_PAGE) {
      // eslint-disable-next-line import/no-cycle
      virtualDetailsPage = (await import(/* webpackChunkName: "virtualDetails" */ './content')).virtualDetailsPage;
    } else if (getState()?.virtualDetails?.pageToDisplay === SESSION_NOT_AVAILABLE_PAGE) {
      // eslint-disable-next-line import/no-cycle
      virtualDetailsPage = (await import(/* webpackChunkName: "virtualDetails" */ './content'))
        .sessionNotAvailableMessagePage;
    } else {
      // eslint-disable-next-line import/no-cycle
      virtualDetailsPage = (await import(/* webpackChunkName: "virtualDetails" */ './content')).genericMessagePage;
    }

    dispatch(injectPageSnapshot(virtualDetailsPage));
  };

export const virtualDetailsReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case SET_VIRTUAL_DETAILS:
    case DISPLAY_SESSION_NOT_AVAILABLE_MESSAGE:
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
