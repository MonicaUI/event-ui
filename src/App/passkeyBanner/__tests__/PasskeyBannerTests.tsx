import React from 'react';
import { combineReducers } from 'redux';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import EventSnapshot from '../../../../fixtures/EventSnapshot.json';
import createStoreWithMiddleware from 'nucleus-guestside-site/src/redux/devCreateStoreWithMiddleware';
import PasskeyBanner from '../PasskeyBanner';

const website = EventSnapshot.eventSnapshot.siteEditor.website;
const registrationPaths = website.pluginData.registrationProcessNavigation.registrationPaths;
const regPathId = Object.getOwnPropertyNames(registrationPaths)[0];
const cancellationPageId = registrationPaths[regPathId].registrationCancellationPageIds[0];

const eventRegId = 'eventRegistrationId1';

const appState = {
  defaultUserSession: {
    isTestMode: false,
    isPlanner: false,
    isPreview: false
  },
  website,
  text: {
    translate: (resx, tokens) => (tokens ? `${resx}:${JSON.stringify(tokens)}` : resx)
  },
  pathInfo: {
    currentPageId: cancellationPageId
  },
  eventTravel: {
    hotelsData: {
      isPasskeyEnabled: true
    }
  },
  travelCart: {
    cart: {
      bookings: [
        {
          id: eventRegId,
          hotelRoomBookings: [
            {
              requestedAction: 'BOOK',
              id: 'hotelRoomBookingId1',
              hotelReservationDetailId: 'hotelReservationDetailId1'
            }
          ]
        }
      ]
    }
  },
  registrationForm: {
    currentEventRegistrationId: eventRegId,
    regCart: {
      eventRegistrations: {
        [eventRegId]: {}
      }
    }
  },
  appData: {
    registrationSettings: {
      registrationPaths: {
        [regPathId]: {}
      }
    }
  }
};

const createStore = initialState => {
  return createStoreWithMiddleware(
    combineReducers({
      userSession: (x = {}) => x,
      defaultUserSession: (x = {}) => x,
      website: (x = {}) => x,
      text: (x = {}) => x,
      pathInfo: (x = {}) => x,
      eventTravel: (x = {}) => x,
      travelCart: (x = {}) => x,
      registrationForm: (x = {}) => x,
      appData: (x = {}) => x
    }),
    initialState
  );
};

describe('Passkey Banner', () => {
  test('should render', () => {
    const store = createStore(appState);
    const page = mount(
      <Provider store={store}>
        <PasskeyBanner />
      </Provider>
    );
    expect(page.find('PasskeyBanner').first()).toMatchSnapshot();
  });

  test('should NOT render when passkey is not enabled', () => {
    const initialState = {
      ...appState,
      eventTravel: {
        ...appState.eventTravel,
        hotelsData: {
          ...appState.eventTravel.hotelsData,
          isPasskeyEnabled: false
        }
      }
    };
    const store = createStore(initialState);
    const page = mount(
      <Provider store={store}>
        <PasskeyBanner />
      </Provider>
    );
    expect(page.find('PasskeyBanner').first()).toMatchSnapshot();
  });

  test('should NOT render when there are no active passkey bookings', () => {
    const initialState = {
      ...appState,
      travelCart: {
        ...appState.travelCart,
        cart: {
          ...appState.travelCart.cart,
          bookings: [
            {
              ...appState.travelCart.cart.bookings[0],
              hotelRoomBookings: [
                {
                  ...appState.travelCart.cart.bookings[0].hotelRoomBookings[0],
                  requestedAction: 'CANCEL'
                }
              ]
            }
          ]
        }
      }
    };
    const store = createStore(initialState);
    const page = mount(
      <Provider store={store}>
        <PasskeyBanner />
      </Provider>
    );
    expect(page.find('PasskeyBanner').first()).toMatchSnapshot();
  });

  test('should NOT render when page is not cancellation page', () => {
    const initialState = {
      ...appState,
      pathInfo: {
        ...appState.pathInfo,
        currentPageId: 'notCancellation'
      }
    };
    const store = createStore(initialState);
    const page = mount(
      <Provider store={store}>
        <PasskeyBanner />
      </Provider>
    );
    expect(page.find('PasskeyBanner').first()).toMatchSnapshot();
  });
});
