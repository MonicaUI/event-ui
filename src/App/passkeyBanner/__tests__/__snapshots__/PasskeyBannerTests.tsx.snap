// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Passkey Banner should NOT render when page is not cancellation page 1`] = `null`;

exports[`Passkey Banner should NOT render when passkey is not enabled 1`] = `null`;

exports[`Passkey Banner should NOT render when there are no active passkey bookings 1`] = `null`;

exports[`Passkey Banner should render 1`] = `
<div>
  <div />
  <div
    class="banner sticky-page-banner"
  >
    <div
      class="wrapper"
      style="background-color: rgb(221, 226, 230);"
    >
      <div
        class="iconWrapper"
        style="color: rgb(4, 21, 50);"
      >
        <span
          class="container"
        >
          <span
            aria-label="_attention__resx"
            class="icon"
            role="img"
          />
        </span>
      </div>
      <div
        class="contentWrapper"
      >
        <h4
          class=""
          style="font-family: PT Sans, sans-serif; font-weight: 400; font-size: 1.375rem; color: rgb(4, 21, 50); line-height: 1.3; font-style: normal; background-color: transparent; border-radius: 0; padding: 0px 0px 0px 0px;"
        >
          EventWidgets_Passkey_RegCancelStickyHeader__resx
        </h4>
        <div
          class=""
          style="font-family: PT Sans, sans-serif; font-weight: 400; font-size: 16px; color: rgb(39, 63, 105); line-height: 1.3; font-style: normal; background-color: transparent; border-radius: 0; padding: 0px 0px 0px 0px;"
        >
          <span>
            EventWidgets_Passkey_RegCancelStickyFirstLine__resx
          </span>
        </div>
        <div
          class="actionBar"
        >
          <button
            class=""
            data-cvent-id="passkey-banner-button-cancel"
            style="font-family: PT Sans, sans-serif; font-weight: 400; font-size: 1.25rem; color: rgb(255, 255, 255); line-height: 1.3; font-style: normal; background-color: rgb(4, 21, 50); border-radius: 6px; border-width: 1px; border-style: solid; border-color: #041532; padding: 5px 1.25rem 5px 1.25rem;"
            title="EventWidgets_Passkey_RegCancelStickyCancelButton__resx"
            type="button"
          >
            <span>
              EventWidgets_Passkey_RegCancelStickyCancelButton__resx
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
